### PR TITLE
Extract WinHttp code into mockable service, add Request/Response data structures

### LINKED ISSUE #114: Passwords are not URL-escaped
I wasn't able to log in to RA through any of the emulator apps, so I changed my password to a simple one. I was able to log in with this simple password. After changing my password back to a more complex one, I wasn't able to log in again. This made me think it was an issue with not escaping special characters on the login request. I skimmed the source code, and saw that passwords are sent as-is without any checks to see if the characters are valid URL characters. Sure enough, after manually escaping my password and pasting the escaped version into RAIntegration's password field, I was able to log in.

For clarification, please consider the following example:

Let's say this string is my password (it's not my real password!): s3?AJhs:u+S=D<%-]@{O

If I use this string as-is in RAIntegration, my login will fail, even though I would be able to log in to RetroAchievements.org with this very password.

However, if I enter an _**escaped**_ version into RAIntegration, my login will succeed without issue. So my above password would become:

s3%3FAJhs%3Au%2BS%3DD%3C%25-%5D%40%7BO

So instead of sending this request:

`https://retroachievements.org/login_app.php?u=ubergeek77&p=s3?AJhs:u+S=D<%-]@{O`

it sends this request:

`https://retroachievements.org/login_app.php?u=ubergeek77&p=s3%3FAJhs%3Au%2BS%3DD%3C%25-%5D%40%7BO`

and the server can properly read the password and send the appropriate response.

RAIntegration should really be doing this manually, as sending _**any**_ URL request before properly sanitizing it is bad practice.

### PR DESCRIPTION
**Pull request by Jamiras**
* `IHttpRequester` contains the actual logic for making the HTTP call
* `Http::Request` defines the parameters to construct the HTTP request
* `Http::Response` contains the response and may be returned directly via `Http::Request::Call` or through an asynchronous callback via `Http::Request::CallAsync`.
* Url parameters are now URL encoded (fixes #114)
* Image downloads are no longer routed through RAWeb. They now use `IHttpRequester::Download`, which writes the image directly to disk without first buffering it in RAM.

### CODE REVIEW

File: src/RA_httpthread.cpp [hunk 1]
```diff
@@ -327,242 +204,49 @@ BOOL RAWeb::DoBlockingRequest(RequestType nType, const PostArgs& PostData, rapid
...
     }
 
-    //  Debug logging...
-    if (ResponseOut.size() > 0)
-    {
-        rapidjson::Document doc;
-        doc.Parse(ResponseOut.c_str());
+    ra::services::Http::Request request(sUrl);
+    request.SetPostData(sPostData);
+    auto response = request.Call();
```
Comment (Jamiras) [inline src/RA_httpthread.cpp]: Illustrates synchronous call

File: src/services/ImageRepository.cpp [hunk 2]
```diff
@@ -164,31 +165,36 @@ void ImageRepository::FetchImage(ImageType nType, const std::string& sName)
...
+    RA_LOG_INFO("Downloading %s", sUrl.c_str());
 
-    ra::services::ServiceLocator::GetMutable<ra::services::IThreadPool>().RunAsync([this, nRequestType, args, sFilename]()
+    Http::Request request(sUrl);
+    request.DownloadAsync(sFilename, [this,sFilename,sUrl](const Http::Response& response)
     {
-        std::string sResponse;
-        if (RAWeb::DoBlockingRequest(nRequestType, args, sResponse))
-            _WriteBufferToFile(sFilename, sResponse);
+        if (response.StatusCode() == 200)
```
Comment (Jamiras) [inline src/services/ImageRepository.cpp]: Illustrates asynchronous download call.

File: src/services/Http.hh [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    class Response
+    {
+    public:
+        explicit Response(int nStatusCode, const std::string& sResponse) noexcept
+            : m_nStatusCode(nStatusCode), m_sResponse(sResponse)
+        {
+        }
+
+        Response(const Response&) noexcept = default;
+        Response(Response&&) noexcept = default;
```
Comment (SyrianBallaS) [inline src/services/Http.hh]: You should probably default the other operations too. Destructor and the assignments. If you didn't want the assignments you should delete them.

File: src/services/Http.hh [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                m_sUrl = sUrl;
+            }
+            else
+            {
+                m_sUrl.assign(sUrl, 0, nIndex);
+                m_sQueryString.assign(sUrl, nIndex + 1, std::string::npos);
+            }
+        }
+
+        Request(const Request&) = default;
```
Comment (SyrianBallaS) [inline src/services/Http.hh]: You should default and/or delete all of them if you're going to do this. It'll also make it clear what's to be used so we don't have to guess what's implicitly deleted. If you want copying, moving can't be deleted.

File: src/services/IHttpRequester.hh [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+#include "services\Http.hh"
+#include "services\TextWriter.hh"
+
+namespace ra {
+namespace services {
+
+class IHttpRequester
+{
+public:
+    virtual ~IHttpRequester() noexcept = default;
```
Comment (SyrianBallaS) [inline src/services/IHttpRequester.hh]: You should delete/default all of them.

**Pull request merged**